lib/utils/token_utilities.dart

```diff
@@ -220,10 +221,77 @@ class TokenUtilities {
     return 1;
   }
 
-  /// Checks if two currencies match (by title and tag)
-  static bool _matchesToken(Erc20Token token, CryptoCurrency currency) {
-    return token.title.toLowerCase() == currency.title.toLowerCase() &&
-        (token.tag?.toLowerCase() == currency.tag?.toLowerCase() ||
-            (token.tag == null && currency.tag == null));
+  static Future<List<CryptoCurrency>> getAvailableTokensForNetwork(
+    WalletType network,
+    Box<WalletInfo> walletInfoSource,
+  ) async {
+    final baseCurrency = walletTypeToCryptoCurrency(network);
+    final allTokens = <CryptoCurrency>[];
+    final addedAddresses = <String>{};
+
+    allTokens.add(baseCurrency);
+
+    for (final currency in CryptoCurrency.all) {
+      final matches = baseCurrency == CryptoCurrency.eth
+          ? baseCurrency.title == currency.tag
+          : currency.tag == baseCurrency.tag;
+
+      if (matches && _shouldAddToken(allTokens, currency, addedAddresses)) {
+        allTokens.add(currency);
+        if (currency is Erc20Token) {
+          addedAddresses.add(currency.contractAddress.toLowerCase());
+        }
+      }
+    }
+
+    // Add user tokens that don't already exist
+    final userTokens = await _getUserTokensForNetwork(baseCurrency, walletInfoSource);
+    for (final token in userTokens) {
+      if (_shouldAddToken(allTokens, token, addedAddresses)) {
+        allTokens.add(token);
+        if (token is Erc20Token) {
+          addedAddresses.add(token.contractAddress.toLowerCase());
+        }
+      }
+    }
+
+    return allTokens;
+  }
+
+  static bool _shouldAddToken(
+    List<CryptoCurrency> existingTokens,
+    CryptoCurrency token,
+    Set<String> addedAddresses,
+  ) {
+    if (token is Erc20Token) {
+      final address = token.contractAddress.toLowerCase();
+      if (addedAddresses.contains(address)) {
+        return false;
+      }
+      if (existingTokens.any((existing) => _matchesCurrency(existing, token))) {
+        return false;
+      }
+      return true;
+    }
+    
+    return !existingTokens.any((existing) => _matchesCurrency(existing, token));
+  }
+
+  static bool _matchesCurrency(CryptoCurrency a, CryptoCurrency b) {
+    return a.title.toUpperCase() == b.title.toUpperCase() &&
+        (a.tag?.toUpperCase() == b.tag?.toUpperCase() || (a.tag == null && b.tag == null));
+  }
+
+  static Future<List<CryptoCurrency>> _getUserTokensForNetwork(
+    CryptoCurrency baseCurrency,
+    Box<WalletInfo> walletInfoSource,
+  ) async {
+    final tokens = await TokenUtilities.loadAllUniqueEvmTokens(walletInfoSource);
+
+    return tokens.where((token) {
+      if (baseCurrency == CryptoCurrency.eth) return token.tag == baseCurrency.title;
+
+      return token.tag?.toLowerCase() == baseCurrency.tag?.toLowerCase();
```

Comment:
same, will miss solana and tron, and will probably match any currencies that has no tags together, although I think this function is only called with EVM, SOL, and TRX so it's safe, but double check it please, or safer, modify this